### PR TITLE
Feature/history api for restore layer

### PR DESCRIPTION
**Pull request by sonic182**
This works but I need to add something to fix this ->

imagen

maybe a configurable callback for applying some styles when searching..?

* [x] Load from url if state not present
* [ ] Setup an callback for applying some styles

### CODE REVIEW

**Comment by carlosescri** [comment]
Your problem is that the layer is displayed "on load" but the search box is not visible so it gets positioned… at `(0, 0)`, right?

**Comment by sonic182** [comment]
> Your problem is that the layer is displayed "on load" but the search box is not visible so it gets positioned… at `(0, 0)`, right?

I think we need to add the same callback of clicking the "search icon" somewhere when restoring a history state

**Comment by sonic182** [comment]
When doing "popstate" from maybe a product view, if query was in a page > 1 doesn't display the layer correctly

imagen

File: packages/core/src/store/context-history.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        const val = query[key];
+
+        if (val === [] || val === {}) {
+          delete query[key];
+        }
+      }
+
+      // TODO: remove this when layer can start from a page after 1
+      // Doesn't load correctly if page > 1
+      query.page = 1;
```
Comment (sonic182) [inline packages/core/src/store/context-history.ts]: This is a workaround for displaying the layer correctly

**Comment by sonic182** [comment]
> Your problem is that the layer is displayed "on load" but the search box is not visible so it gets positioned… at `(0, 0)`, right?

Resumed answerd: Yes

File: packages/core/src/store/context-history.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      // allow save state for new search's
+      this._skipNextSearch = false;
+    }
+  }
+
+  private _setState(query: Record<string, any>): void {
+    if (query) {
+      // TODO: remove this when layer can start from a page after 1
+      // Doesn't load correctly if page > 1
+      query.page = 1;
```
Comment (sonic182) [inline packages/core/src/store/context-history.ts]: This is a workaround for displaying the layer correctly

**Comment by carlosescri** [comment]
@sonic182 how do you manage multiple contexts that can be restored?

**Comment by sonic182** [comment]
> @sonic182 how do you manage multiple contexts that can be restored? Well, I see that you only apply history to the `default` context, but there could be multiple default contexts (each hashid has a default context).

I think that won't happen, multiple search layers in a web view doesn't sound to me a thing we will see.

Now only default context has the "history restore", and it is subject to changes. Maybe we could allow the customers to add a flag/to mark a context so internally if we see that mark, apply the ContextHistory

**Comment by carlosescri** [comment]
For instance, you could have a _classic_ layer and an _embedded_ layer in the same page at the same time, with different context names (you could have `popup` and `embedded` contexts for the same hashid).

The serialized data should reflect the context that saved it. I would have a single history manager, maybe in the store…

When de-serializing, the context name and hashid could be used to get the context and perform the query.

There's only one history in the window, so we could have one single manager that know how to deal with history and contexts…

```
ContextHistory.serialize(context);
ContextManager.deserialize(serializedData); // would use Store.getContext() internally.
```

It's an idea and requires more thinking, but I think this code is too tied to the default context.

**Comment by sonic182** [comment]
@carlosescri The code is tied to `a context` not just to the `default context`, and yes, in the Store could be a logic to apply ContextHistory to all context's which requires them.

The only context's I think doesn't need ConexHistory are the Suggestion's context (that will be eliminated) so yes, could be applied to all context I think

**Comment by carlosescri** [comment]
Replaced by #248

**Pull request closed without merging**